client/tracing/src/logging/mod.rs

```diff
@@ -0,0 +1,531 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate logging library.
+//!
+//! This crate uses tokio's [tracing](https://github.com/tokio-rs/tracing/) library for logging.
+
+#![warn(missing_docs)]
+
+mod directives;
+mod event_format;
+mod layers;
+
+pub use directives::*;
+pub use sc_tracing_proc_macro::*;
+
+use sc_telemetry::{ExtTransport, TelemetryWorker};
+use std::io;
+use tracing::Subscriber;
+use tracing_subscriber::{
+	fmt::time::ChronoLocal,
+	fmt::{
+		format, FormatEvent, FormatFields, Formatter, Layer as FmtLayer, MakeWriter,
+		SubscriberBuilder,
+	},
+	layer::{self, SubscriberExt},
+	registry::LookupSpan,
+	EnvFilter, FmtSubscriber, Layer, Registry,
+};
+
+pub use event_format::*;
+pub use layers::*;
+
+/// Logging Result typedef.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Logging errors.
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum Error {
+	#[error(transparent)]
+	IoError(#[from] io::Error),
+
+	#[error(transparent)]
+	SetGlobalDefaultError(#[from] tracing::subscriber::SetGlobalDefaultError),
+
+	#[error(transparent)]
+	DirectiveParseError(#[from] tracing_subscriber::filter::ParseError),
+
+	#[error(transparent)]
+	SetLoggerError(#[from] tracing_log::log_tracer::SetLoggerError),
+}
+
+macro_rules! disable_log_reloading {
+	($builder:expr) => {{
+		let builder = $builder.with_filter_reloading();
+		let handle = builder.reload_handle();
+		set_reload_handle(handle);
+		builder
+	}};
+}
+
+/// Common implementation to get the subscriber.
+fn get_subscriber_internal<N, E, F, W>(
```

Comment:
I think this can be named `prepare_subscriber`. It's not `pub` so we know it's internal yes?

Reply:
I want to be sure it's clear on this one :grin: No it's just that originally there was no builder but a bunch of pub functions with a similar name based on `get_subscriber`. I can rename to `prepare_subscriber`, makes more sense indeed.